### PR TITLE
Fetch quartos

### PR DESCRIPTION
**Pull request by erikvatt**

### CODE REVIEW

File: pkg/gcs/gcs.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	for {
+		o, err := objs.Next()
+		if err == iterator.Done {
+			if page == "" {
+				return "", fmt.Errorf("could not find html for id %v", qID)
+			}
+			return page, nil
+		}
+
+		if strings.HasSuffix(o.Name, "/index.html") {
```
Comment (MengZhuNAV) [inline pkg/gcs/gcs.go:66]: Should we also take care of different letter cases? e.g. Index.html

File: pkg/gcs/gcs.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	return datab, nil
+}
+
+func findIndexPage(qID string, objs *storage.ObjectIterator) (string, error) {
+	page := ""
+	for {
+		o, err := objs.Next()
+		if err == iterator.Done {
```
Comment (MengZhuNAV) [inline pkg/gcs/gcs.go:59]: What will happen if we got an error other than iterator.Done? should we log the error? 
Is it possible that an error can prevent the iterator from continuing?

File: pkg/graph/quarto.go [hunk 3]
```diff
@@ -12,7 +12,8 @@ import (
 )
 
 func WriteFileToBucket(ctx context.Context, quartoStoryID string,
-	file graphql.Upload) error {
+	file graphql.Upload,
```
Comment (MengZhuNAV) [inline pkg/graph/quarto.go:15]: Maybe we should put the writing code to the client as well, but this can be done later

File: pkg/quarto/http.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		w.Header().Add("content-type", "application/json")
+	}
+
+	w.Write(objBytes)
+}
+
+func (h *Handler) QuartoMiddleware(next http.Handler) http.Handler {
+	regex, _ := regexp.Compile(`[\n]*\.[\n]*`) // check if object path has file extension
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !regex.MatchString(r.URL.Path) {
```
Comment (MengZhuNAV) [inline pkg/quarto/http.go:65]: I dont quite understand this code: if the path does not match the pattern of a file name with extension, we will redirect and get the object for the user? It seems we should do so when the path matches?

File: pkg/gcs/gcs.go (hunk 1 above)
Comment (erikvatt) [reply pkg/gcs/gcs.go]: Agree, I added a tolower conversion on the string before checking

File: pkg/gcs/gcs.go (hunk 2 above)
Comment (erikvatt) [reply pkg/gcs/gcs.go:62]: From doc:
>In addition, if Next returns an error other than iterator.Done, all subsequent calls will return the same error. To continue iteration, a new `ObjectIterator` must be created.

I have changed it now to log the error and return not found to the user, do you agree?

File: pkg/graph/quarto.go (hunk 3 above)
Comment (erikvatt) [reply pkg/graph/quarto.go:15]: I agree

File: pkg/gcs/gcs.go (hunk 2 above)
Comment (MengZhuNAV) [reply pkg/gcs/gcs.go:62]: Yes, as long as you return on error, we could avoid deadloop and it should be ok

File: pkg/quarto/http.go (hunk 4 above)
Comment (erikvatt) [reply pkg/quarto/http.go:65]: I'm open for suggestions on how to solve this.

What I wanted to achieve was that when accessing e.g. /quarto/<id> we get redirected to the index.html file (or some other .html file should index not exist) for that quarto id.

So when attempting to access a file (e.g. .js, .css, .html etc.) the middleware shall ensure that the GetObject handler is called. If it does not match a file it shall search for the index file using the path prefix and redirect.

Again, I'm open for better alternatives on to how solve this, but I'm not quite sure what you mean when you're saying we should redirect when the path matches?

**Review by MengZhuNAV — APPROVED** [review]

**Pull request merged**